Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+block is the diagonal matrix `svdσ` above. -/
+noncomputable def svdS (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix ((Fin A.rank) ⊕ (Fin (N - A.rank))) ((Fin A.rank) ⊕ (Fin (N - A.rank))) ℝ :=
+  (reindex (eigenColumnEquiv A) (eigenColumnEquiv A))
+    (diagonal (isHermitian_transpose_mul_self A).eigenvalues)
+
+/-- Given a matrix A of size m × n: `svdS'` is a matrix of the same dimensions but partitioned into
+four blocks such that S₁₁ contains the non-zero singular values, S₁₂, S₂₁ and S₂₂ are zeros. The S₁₁
+block is the diagonal matrix `svdσ` above but permuted by the same arbitrary order relating svdU₁
+and svdU₁' above . -/
+noncomputable def svdS' (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Matrix ((Fin A.rank) ⊕ (Fin (M - A.rank))) ((Fin A.rank) ⊕ (Fin (M - A.rank))) ℝ :=
+  (reindex (eigenRowEquiv A) (eigenRowEquiv A))
+    (diagonal (isHermitian_mul_conjTranspose_self A).eigenvalues)
+
+lemma U_columns' (A : Matrix (Fin M) (Fin N) 𝕂) :
+  ((reindex (Equiv.refl (Fin M)) (eigenRowEquiv A))
+    (isHermitian_mul_conjTranspose_self A).eigenvectorMatrix) = fromColumns A.svdU₁' A.svdU₂ := by
+  rw [svdU₂, svdU₁']
+  simp only [reindex_apply, Equiv.refl_symm, Equiv.coe_refl, fromColumns_toColumns]
+
+lemma V_conjTranspose_mul_V (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (A.svdV₁ᴴ  *  A.svdV₁ = 1 ∧ A.svdV₂ᴴ  *  A.svdV₂ = 1) ∧
+    (A.svdV₂ᴴ  *  A.svdV₁ = 0 ∧ A.svdV₁ᴴ  *  A.svdV₂ = 0) := by
+  simp_rw [svdV₁, toColumns₁, svdV₂, toColumns₂, reindex_apply, Equiv.refl_symm, Equiv.coe_refl,
+    submatrix_apply, id_eq, HMul.hMul, dotProduct, conjTranspose_apply, of_apply,
+    ← conjTranspose_apply, IsHermitian.conjTranspose_eigenvectorMatrix, ← mul_apply,
+    Matrix.mul_eq_one_comm.1 (IsHermitian.eigenvectorMatrix_mul_inv _)]
+  constructor
+  swap
+  simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, EmbeddingLike.apply_eq_iff_eq,
+   not_false_eq_true, one_apply_ne, zero_apply]
+  exact ⟨rfl, rfl⟩
+  constructor
+  all_goals (
+    funext i j
+    by_cases i = j
+    simp_rw [h, one_apply_eq, one_apply_ne h]
+    rw [one_apply_ne]
+    simp only [ne_eq, EmbeddingLike.apply_eq_iff_eq, Sum.inr.injEq, Sum.inl.injEq, h]
+  )
+
+lemma V₁_conjTranspose_mul_V₁ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdV₁ᴴ * A.svdV₁ = 1 :=
+  (V_conjTranspose_mul_V A).1.1
+
+lemma V₂_conjTranspose_mul_V₂ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdV₂ᴴ * A.svdV₂ = 1 :=
+  (V_conjTranspose_mul_V A).1.2
+
+lemma V₂_conjTranspose_mul_V₁ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdV₂ᴴ * A.svdV₁ = 0 :=
+  (V_conjTranspose_mul_V A).2.1
+
+lemma V₁_conjTranspose_mul_V₂ (A : Matrix (Fin M) (Fin N) 𝕂) : A.svdV₁ᴴ * A.svdV₂ = 0 :=
+  (V_conjTranspose_mul_V A).2.2
+
+lemma V_inv (A : Matrix (Fin M) (Fin N) 𝕂) :
+    (fromColumns A.svdV₁ A.svdV₂)ᴴ  *  (fromColumns A.svdV₁ A.svdV₂) = 1 := by
+  rw [conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromRows_mul_fromColumns,
+    V₁_conjTranspose_mul_V₂, V₁_conjTranspose_mul_V₁, V₂_conjTranspose_mul_V₂,
+    V₂_conjTranspose_mul_V₁, fromBlocks_one]
```

Comment:
I suppose this is the statement that `V` is unitary. Could you add a similar result for `V * Vᴴ`?

Reply:
OK